File: tests/npm/test_2_2_1_multi_property_support.ts
/**
 * Tests for Task 2.2.1: Multi-property Support.
 * Mirrors tests/python/test_2_2_1_multi_property_support.py
 */
import { describe, it, expect, beforeEach } from 'vitest'
import tc from './tc'
import type { TinkerCat, Vertex, Edge } from './tinkercat.d.ts'

describe('Multi-property Support', () => {
  let graph: TinkerCat

  beforeEach(() => { graph = tc.createTinkerCat() })

  it('stores string property', () => {
    const v: Vertex = tc.addVertexWithLabel(graph, 'item', { name: 'hello' })
    expect(tc.getPropertyValue(v, 'name')).toBe('hello')
  })

  it('stores integer property', () => {
    const v: Vertex = tc.addVertexWithLabel(graph, 'item', { count: 42 })
    expect(tc.getPropertyValue(v, 'count')).toBe(42)
  })

  it('stores float property', () => {
    const v: Vertex = tc.addVertexWithLabel(graph, 'item', { score: 3.14 })
    expect(tc.getPropertyValue(v, 'score')).toBeCloseTo(3.14)
  })

  it('stores boolean property', () => {
    const v: Vertex = tc.addVertexWithLabel(graph, 'item', { active: true })
    expect(tc.getPropertyValue(v, 'active')).toBe(true)
  })

  it('stores multiple properties', () => {
    const v: Vertex = tc.addVertexWithLabel(graph, 'person', { name: 'Alice', age: 25, active: true })
    expect(tc.getPropertyValue(v, 'name')).toBe('Alice')
    expect(tc.getPropertyValue(v, 'age')).toBe(25)
    expect(tc.getPropertyValue(v, 'active')).toBe(true)
  })

  it('getPropertyKeys returns all stored keys', () => {
    const v: Vertex = tc.addVertexWithLabel(graph, 'person', { name: 'Bob', age: 30 })
    const keys: string[] = tc.getPropertyKeys(v)
    expect(keys).toContain('name')
    expect(keys).toContain('age')
  })

  it('property can be updated via setProperty', () => {
    const v: Vertex = tc.addVertexWithLabel(graph, 'item', { count: 1 })
    tc.setProperty(v, 'count', 99)
    expect(tc.getPropertyValue(v, 'count')).toBe(99)
  })

  it('edge stores properties', () => {
    const a: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'A' })
    const b: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'B' })
    const e: Edge = tc.addEdge(a, 'link', b, { weight: 0.9 })
    expect(tc.getPropertyValue(e, 'weight')).toBeCloseTo(0.9)
  })

  it('edge label is correct', () => {
    const a: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'A' })
    const b: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'B' })
    const e: Edge = tc.addEdge(a, 'knows', b, {})
    expect(e.label()).toBe('knows')
  })

  it('vertex id is unique', () => {
    const a: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'A' })
    const b: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'B' })
    expect(a.id()).not.toBe(b.id())
  })

  it('property keys are case-sensitive', () => {
    const v: Vertex = tc.addVertexWithLabel(graph, 'item', { Name: 'Alice' })
    expect(tc.hasProperty(v, 'Name')).toBe(true)
    expect(tc.hasProperty(v, 'name')).toBe(false)
  })

  it('vertex without label gets default label', () => {
    const v: Vertex = tc.addVertex(graph, { name: 'NoLabel' })
    expect(v.label()).toBeDefined()
  })

  it('large number of vertices is handled', () => {
    for (let i = 0; i < 100; i++) {
      tc.addVertexWithLabel(graph, 'item', { index: i })
    }
    expect(tc.getVertexCount(graph)).toBe(100)
  })
})

File: tests/kotlin/test_2_2_1_multi_property_support.kt
package org.apache.tinkerpop.gremlin.tinkercat.tests

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerCat
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerVertex

/**
 * Tests for Task 2.2.1: Multi-Property Support.
 *
 * Validates that vertices and edges can carry multiple properties of various
 * types, that properties can be read back without loss, and that absent
 * properties return null.
 * Mirrors tests/python/test_2_2_1_multi_property_support.py.
 */
class Test_2_2_1_MultiPropertySupport : StringSpec({

    lateinit var graph: TinkerCat

    beforeTest { graph = TinkerCat.open() }

    "vertex can store a string property" {
        val v = graph.addVertex(TinkerVertex.T.label, "item", "name", "Widget")
        v.value<String>("name") shouldBe "Widget"
    }

    "vertex can store an integer property" {
        val v = graph.addVertex(TinkerVertex.T.label, "item", "count", 42)
        v.value<Int>("count") shouldBe 42
    }

    "vertex can store a long property" {
        val v = graph.addVertex(TinkerVertex.T.label, "item", "big", 9_000_000_000L)
        v.value<Long>("big") shouldBe 9_000_000_000L
    }

    "vertex can store a double property" {
        val v = graph.addVertex(TinkerVertex.T.label, "item", "price", 3.14)
        v.value<Double>("price") shouldBe 3.14
    }

    "vertex can store a boolean property" {
        val v = graph.addVertex(TinkerVertex.T.label, "flag", "active", true)
        v.value<Boolean>("active") shouldBe true
    }

    "vertex can carry multiple properties simultaneously" {
        val v = graph.addVertex(
            TinkerVertex.T.label, "person",
            "name", "Alice",
            "age", 30,
            "active", true,
            "score", 9.5
        )
        v.value<String>("name") shouldBe "Alice"
        v.value<Int>("age") shouldBe 30
        v.value<Boolean>("active") shouldBe true
        v.value<Double>("score") shouldBe 9.5
    }

    "absent property returns null" {
        val v = graph.addVertex(TinkerVertex.T.label, "item", "name", "X")
        v.value<String>("missing") shouldBe null
    }

    "property set via property() is readable via value()" {
        val v = graph.addVertex(TinkerVertex.T.label, "item") as TinkerVertex
        v.property("score", 99)
        v.value<Int>("score") shouldBe 99
    }

    "property update overwrites previous value" {
        val v = graph.addVertex(TinkerVertex.T.label, "item") as TinkerVertex
        v.property("score", 10)
        v.property("score", 20)
        v.value<Int>("score") shouldBe 20
    }

    "edge can carry properties" {
        val a = graph.addVertex(TinkerVertex.T.label, "n")
        val b = graph.addVertex(TinkerVertex.T.label, "n")
        val e = a.addEdge("knows", b, "since", 2020, "weight", 0.8)
        e.value<Int>("since") shouldBe 2020
        e.value<Double>("weight") shouldBe 0.8
    }

    "edge property set separately is readable" {
        val a = graph.addVertex(TinkerVertex.T.label, "n")
        val b = graph.addVertex(TinkerVertex.T.label, "n")
        val e = a.addEdge("link", b)
        e.property("tag", "important")
        e.value<String>("tag") shouldBe "important"
    }

    "distinct vertices with same property key are independent" {
        val v1 = graph.addVertex(TinkerVertex.T.label, "item", "x", 1)
        val v2 = graph.addVertex(TinkerVertex.T.label, "item", "x", 2)
        v1.value<Int>("x") shouldBe 1
        v2.value<Int>("x") shouldBe 2
    }

    "property key is case-sensitive" {
        val v = graph.addVertex(TinkerVertex.T.label, "item", "Name", "Alice")
        v.value<String>("Name") shouldBe "Alice"
        v.value<String>("name") shouldBe null
    }
})
